### PR TITLE
Assembler MVP

### PR DESCRIPTION
**Pull request by CharlesAverill**
## Features
- Basic MIPS assembler:
  - `add`, `sub`, `sll`, `slr`, `xor`, `lui` instructions
  - All registers
  - Identical output to `mips-linux-gnu-as`
  - Example input/output in `.artifacts`
- Command line argument parsing:
  - Only recognizes positional, required arguments (no options)
  - Help menu
- Config file parsing:
  - Named configurations
  - All values required due to `toml` parser
  - `as_cmd` field allows for custom assembler commands to be passed in (running commands not yet functional)
- Unsigned 5/6-bit integer structs
- Makefile for ease of building

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -1,3 +1,27 @@
...
+pub mod args;
+pub mod config;
+
+use args::parse_args;
+use config::parse_config;
+use nma::nma::assemble;
+
+fn main() -> Result<(), &'static str> {
+    let cmd_args = parse_args()?;
+    let config = parse_config(&cmd_args).expect("Failed to parse configuration file");
```
Comment (qwe-q) [inline src/main.rs:17]: ``.expect()`` immediately kills the program if an error occurs, let's not do that for error-prone filesystem access

File: src/nma/nma.rs [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+        "lui" => Ok(I {
+            opcode: U6 { value: 0xf },
+            form: I_form::RT_IMM,
+        }),
+        _ => Err("Failed to match I-instr mnemonic"),
+    }
+}
+
+pub fn read_file(file_path: &str) -> String {
+    fs::read_to_string(file_path).expect("Failed to read the file")
```
Comment (qwe-q) [inline src/nma/nma.rs:80]: ``.expect()`` immediately kills the program if an error occurs, let's not do that for error-prone filesystem access

Comment (CharlesAverill) [reply src/nma/nma.rs:80]: What should the correct course of action be here? Should I propagate a `Result` back up to `main`? For the other one I can just default to the existing assembler, although in the future I imagine the config file will become important enough that we should expect it

Comment (qwe-q) [reply src/nma/nma.rs:80]: Graceful fatal termination is the correct course of action IMO, the problem is that ``.expect()`` is not graceful and kills the program looking like this:

```
thread 'main' panicked at 'expect error message: "some message"', libcore/result.rs:945:5
```

That's ugly to users even though you are able to provide your own error message.

Comment (CharlesAverill) [reply src/nma/nma.rs]: Resolved

**Review by hkwheat — APPROVED** [review]
looks good to me

**Review by lukewellsdev — APPROVED** [review]
Very coo!l ! !!!!

File: src/nma/nma.rs [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+            } else {
+                Err("Register out of bounds")
+            }
+        }
+    }
+}
+
+/// Assembles an R-type instruction
+fn assemble_r(r_struct: &mut R, r_args: Vec<&str>) -> Result<u32, &'static str> {
+    let mut rs: u32 = 0;
```
Comment (qwe-q) [inline src/nma/nma.rs:185]: You wrote a U5 and U6 type with a complete implementation, and then used u32s for the registers?

File: src/nma/nma.rs [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+            Some(digit) => Ok(digit),
+            _ => Err("Invalid register index"),
+        },
+        _ => Err("Malformed mnemonic"),
+    }
+}
+
+/// Given a register or number, assemble it into its integer representation
+fn assemble_reg(mnemonic: &str) -> Result<u32, &'static str> {
+    if !mnemonic.starts_with("$") {
```
Comment (qwe-q) [inline src/nma/nma.rs:142]: Clippy warning— use a char instead

File: src/nma/nma.rs [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+        }
+    }
+}
+
+/// Assembles an R-type instruction
+fn assemble_r(r_struct: &mut R, r_args: Vec<&str>) -> Result<u32, &'static str> {
+    let mut rs: u32 = 0;
+    let mut rt: u32 = 0;
+    let mut rd: u32 = 0;
+    let mut shamt: u32 = 0;
```
Comment (qwe-q) [inline src/nma/nma.rs:188]: Clippy warning— you don't need to initialize this

File: src/nma/nma.rs [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+        result,
+        shortwidth = 8,
+        width = 32
+    );
+    Ok(result)
+}
+
+/// Assembles an I-type instruction
+fn assemble_i(i_struct: &mut I, i_args: Vec<&str>) -> Result<u32, &'static str> {
+    let mut rs: u32 = 0;
```
Comment (qwe-q) [inline src/nma/nma.rs:248]: Clippy warning— you don't need to initialize this

File: src/nma/nma.rs [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+    Ok(result)
+}
+
+// General assembler entrypoint
+pub fn assemble(args: &Args) -> Result<(), &'static str> {
+    let input_fn = &args.input_as;
+    let output_fn = &args.output_as;
+
+    let file_contents: String;
+    match fs::read_to_string(input_fn) {
```
Comment (qwe-q) [inline src/nma/nma.rs:299]: Clippy warnings about using match initialization instead

File: src/nma/nma.rs [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+    };
+    let mut r_args: Vec<&str> = Vec::new();
+    let mut i_struct: I = I {
+        opcode: U6 { value: 0 },
+        form: IForm::None,
+    };
+    let mut i_args: Vec<&str> = Vec::new();
+
+    // Iterate over all tokens
+    while tokens.len() > 0 {
```
Comment (qwe-q) [inline src/nma/nma.rs:326]: Clippy warning

File: src/nma/nma.rs [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+                    println!(
+                        "[R] {} - shamt [{:x}] - funct [{:x}]",
+                        token, instr_info.shamt, instr_info.funct
+                    );
+
+                    r_struct = instr_info;
+                    r_args.clear();
+                    r_args.push(token)
+                }
+                _ => match i_operation(&token) {
```
Comment (qwe-q) [inline src/nma/nma.rs:349]: Just start listening to Clippy, I'm not gonna spell out the rest of them

File: src/nma/nma.rs [hunk 10]
```diff
@@ -0,0 +1,412 @@
...
+            opcode: U6 { value: 0xf },
+            form: IForm::RtImm,
+        }),
+        _ => Err("Failed to match I-instr mnemonic"),
+    }
+}
+
+/// Split a string into meaningful, atomic elements of the MIPS language
+pub fn tokenize(raw_text: &str) -> Vec<&str> {
+    raw_text.split_whitespace().collect::<Vec<&str>>()
```
Comment (qwe-q) [inline src/nma/nma.rs:90]: Needs better matching of delimiters for greater ergonomics while writing code (this one is a big deal)

File: src/nma/nma.rs [hunk 11]
```diff
@@ -0,0 +1,412 @@
...
+}
+
+/// Write a u32 into a file, zero-padded to 32 bits (4 bytes)
+pub fn write_u32(mut file: &File, data: u32) -> std::io::Result<()> {
```
Comment (qwe-q) [inline src/nma/nma.rs:94]: wait, why is this function necessary at all?

File: src/nma/nma.rs [hunk 12]
```diff
@@ -0,0 +1,412 @@
...
+
+    // Iterate over all tokens
+    while tokens.len() > 0 {
+        let token = tokens.remove(0);
+
+        // Scan tokens in
+        match state {
+            AssemblerState::Initial => match token {
+                "main:" => state = AssemblerState::Scanning,
+                _ => return Err("Code must begin with 'main' label"),
```
Comment (qwe-q) [inline src/nma/nma.rs:333]: does it really, though? Especially if you want to, say, write other labels for functions before your main function. I'd accept this only if it's planned to be expanded on later

**Review by qwe-q — CHANGES_REQUESTED** [review]
Most of what I have to say is "please listen to Clippy" or nitpicks, but I do have a large concern that ergonomics about how people write assembly is not being considered at the earliest stages of the project.

Ask me about my at-scale playtesting idea for NAME.

File: src/nma/nma.rs [hunk 13]
```diff
@@ -0,0 +1,412 @@
...
+        },
+        _ => Err("Malformed mnemonic"),
+    }
+}
+
+/// Given a register or number, assemble it into its integer representation
+fn assemble_reg(mnemonic: &str) -> Result<u32, &'static str> {
+    if !mnemonic.starts_with("$") {
+        return match mnemonic.parse::<u32>() {
+            Ok(v) => Ok(v),
```
[the remark this answers is not in the document]
Comment (qwe-q) [reply src/nma/nma.rs:144]: To be clear, that's a consequence of parsing any valid u32 for registers. I'm sure it'd just get masked off to register 31, but it's a bug.

File: src/nma/nma.rs (hunk 12 above)
Comment (CharlesAverill) [reply src/nma/nma.rs:333]: This is just boilerplate atm, when I get a MIPS grammar written this will go away

File: src/nma/nma.rs (hunk 10 above)
Comment (CharlesAverill) [reply src/nma/nma.rs]: Accounted for all whitespace and comma

Comment (CharlesAverill) [reply src/nma/nma.rs]: This does allow for statements like `add $t0 $t1 $t2` but I'm also going to resolve that with the real parser later

**Comment by CharlesAverill** [comment]
@qwe-q All feedback resolved

**Review by qwe-q — APPROVED** [review]

**Pull request merged**